### DISCUSSION TITLE
No data is coming after loading calibration

### DISCUSSION THREAD

**Discussion by iliavits** [jankae/LibreVNA#256] [General]
Hello,
I have a simple Python program to read the data from LibreVNA. I can upload the data before the calibration, but right after the calbration no data is coming. I only get the "# GHZ S RI R 50.000000000000". In order to bring the data again I need to reburn the firmware. I am uploading the calibration externally from the premeasured file with LibreVNA-GUI.
Thank you.

Here is the program I use:

import socket
import time

# Create a socket
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

# Connect to the LibreVNA
HOST = '127.0.0.1'  # local host
PORT = 19542  # LibreVNA port
s.connect((HOST, PORT))

# Function to send SCPI commands
def send_scpi_command(cmd):
    s.sendall((cmd + '\n').encode())
    time.sleep(0.12)  # Wait for the command to be processed

# Function to query SCPI commands
def query_scpi_command(cmd):
    send_scpi_command(cmd)
    return s.recv(16384).decode()

# Set up the sweep parameters
send_scpi_command(":DEV:MODE VNA")
send_scpi_command(":VNA:SWEEP FREQUENCY")
send_scpi_command(":VNA:ACQ:IFBW 10000")
send_scpi_command(":VNA:ACQ:AVG 1")
send_scpi_command(":VNA:ACQ:POINTS 201")
send_scpi_command(":VNA:FREQuency:START 4000000000")
send_scpi_command(":VNA:FREQuency:STOP 6000000000")

# Sweeps loop
for n in range(4):
    send_scpi_command(":VNA:ACQuisition:SINGLE TRUE")

    # Wait for the sweep to finish
    while query_scpi_command(":VNA:ACQ:FIN?") == "FALSE":
        time.sleep(0.1)

    # Grab the data of trace S11
    npts = int(query_scpi_command(":VNA:ACQ:POINTS?")) + 1
    tmp = query_scpi_command(":VNA:TRACE:TOUCHSTONE? S11")
    print(npts)
    print(tmp)

# Close the socket
s.close()

**Comment by jankae** [comment]
This sounds confusing and I can not quite follow it.

> I can upload the data before the calibration

Not sure what you mean by "upload". Do you mean you want to retrieve the data from the LibreVNA? And when is this calibration applied? I don't see anything related to the calibration in your program. Do you run the program once before loading the calibration (manually through the GUI) and everything works but when you run your program again after loading the calibration you do not get the data?

> I only get the "# GHZ S RI R 50.000000000000"

Did you check these issues related to the touchstone command? #101 and #108

Do you see data in the graphs of the GUI after loading the calibration?

>  In order to bring the data again I need to reburn the firmware.

I find that hard to believe. There is nothing being stored persistently on the device that is related at all to the calibration. Even if you run into issues with loading the calibration, you should be able to power cycle the LibreVNA and restart the GUI to get back to the same state you had before loading the calibration. Are you sure that the problem only reverts once you flash the firmware again?

**Reply by iliavits** [reply]
Yes, by upload I mean retrieve. I have tried now with ":VNA:TRACE:DATA? S11" and after the "# GHZ S RI R 50.000000000000" it returns EMPTY.
I apply the calibration separately through the LibreVNA-GUI.
I am not sure that burning is what is helping, maybe it is just a reset...

**Reply by jankae** [reply]
Can you please also answer my other questions:
1. Are you able to get complete trace data before you apply the calibration with your program?
2. Do you still see the traces updating in the GUI after you applied the calibration

> I have tried now with ":VNA:TRACE:DATA? S11" and after the "# GHZ S RI R 50.000000000000" it returns EMPTY.

Either something is going very wrong or you are not giving me precise information. The `VNA:TRACE:DATA?` query does not return a touchstone file, so its response will not start with `# GHZ S RI R 50.000000000000`. This command will return the trace data as a list of frequency and real/imag values instead.

If you are getting `# GHZ S RI R 50.000000000000` I suspect that you are using the `VNA:TRACE:TOUCHSTONE?` instead (as in your original post). Did you check the other issues for that command? It returns the Touchstone file as it is, including newlines. If your socket receive function returns after the first line, you will not get any of the data in the touchstone file (but the GUI will still have sent it and it is queued in your socket). See #101 and #108 for a more detailed explanation.

I realize that I may come across a bit standoffish in my replies and I am sorry about that. But helping you when you ignore half my questions or are not exactly describing what you area doing just becomes exponentially more difficult.

**Comment by iliavits** [comment]
Thank you for helping. I am confused by myself, so probably my answers are not fully aligned. 
At the beggining I was able to get the data and was happy. Then I am not sure after what I started to get only the touchstone heared without any data. I was desperate and tried to reburn the FW. It seemed to help and Iwas getting data again. Then I maid a calibration procedure with LibreVNA-GUI and tried again to get the data with my Python program. I didn't get any data except the header. When I tried to get the data with ":VNA:TRACE:DATA? S11" I only got EMPTY (no touchstone header - my mistake).
When I try to use LibreVNA-GUI  I can see traces updating.
I appologize for my reporting, I am rather newby to the VNA, so many concepts are not fully clear to me.

**Comment by jankae** [comment]
Okay, but can you describe exactly what you are doing? My understanding at the moment is this:

1. You started the GUI and connected the LibreVNA
2. You are running your script and manage to read touchstone data
3. You perform a calibration through the GUI manually
4. The GUI shows the calibrated traces
5. You run the same script again but this time you only get the Touchstone header, no data

Not sure if this is all correct. After this, I am getting really confused:

6. You updated the firmware on the LibreVNA
7. You run the same script again and are getting data now

I am sure that at least some steps are missing or I understood them wrong. Especially the update fixing the issue makes no sense at all. So please write me a list like this, listing every single action you perform in the GUI and attach your program. I must be able to replicate the exact same procedure without having to guess what you did exactly or in what order.

**Reply by iliavits** [reply]
Steps 1-5 are correct, then the steps 6 and 7 are confusing for me too. I am not sure about those steps since they seemed to work one time and now they seem to get confused results, sometimes working sometimes not.
I suggest that we do a reset to the device and repeat steps from 1 to 5 and see if we will get the data or not.
So how do I do the full reset to the device?

**Reply by iliavits** [reply]
Ok, I did the following:
0. Disconnet the power and USB from LibreVNA.
0a. Reconnect the USB and Power to LibreVNA.
1. Run LibreVNA-GUI - I can get the data and se the traces changing.
2. Disconect the device in LibreVNA-GUI to be able to work with my script.
3. Run Python script to get some data from VNA. The program succed to report the number of points that I requested with ":VNA:ACQ:POINTS?" command, but got an "EMPTY" string when I requested ":VNA:TRACE:DATA? S11"

**Comment by jankae** [comment]
>  then the steps 6 and 7 are confusing for me too

How can they confuse you, you are the one performing them? I am not talking about confusing results, that can obviously happen. But you must know what you are doing.

So can you please write a list (in your own) words what exactly you are doing and what results you are getting from your program? Broken down into as many details as you can. I'd be happy with a description of every single button press or mouse click if necessary. Alternatively, a screen capture video might also work. I am absolutely still not sure what you are doing and I also do not know how I can make it any more clear that your description so far is not sufficient for me to understand it. And if I don't know the precise steps, I can't even start to think about helping with the problem.

You seem to be able to repeat the problem at least some of the time. This is good and I would like to do the same.

> So how do I do the full reset to the device?

Power cycle the LibreVNA and restart the GUI.

**Comment by iliavits** [comment]
Thanks, I am confused in the sense that I am not sure what exact steps I did after steps 1-5. I played with it a lot and can't be sure anymore. Therefore, I will repower the device and repeat steps 1-5 trying to recreate the problem. I will report you then.
Thanks, again.

**Comment by jankae** [comment]
Answer to [this](https://github.com/jankae/LibreVNA/discussions/256#discussioncomment-9594709) reply:

Perfect, I can reproduce that as well. I got confused earlier because I asked "Do you see data in the graphs of the GUI after loading the calibration?" and the only thing I could interpret as a reply to this was "When I try to use LibreVNA-GUI I can see traces updating." Which is not true at the time when your script does not return data anymore.

Also, nothing in your steps refers to any calibration which you mentioned in the original post and which also really confused me. If you had provided this list of steps from the beginning, it would have been a really easy thing to solve.

Anyway, lets get into the issue. Here is what I see when I replicate your steps (and yes, this is the level of detail that I would ideally like from you):

> 1. Run LibreVNA-GUI - I can get the data and se the traces changing.

- Start the GUI with a device connected. The LibreVNA is capturing data and all four graphs are updating with new data
- Run the script (as given in your original post)
- While the script is running, all traces in the GUI get cleared and then a new measurement is taken (new traces appear) four times
- After running your script, the traces in the GUI are still visible but not updating anymore
- The output of the script shows the trace data (repeated four times, once for each sweep)

> 2. Disconect the device in LibreVNA-GUI to be able to work with my script.

- Select Device->Disconnect in the menu
- The traces in the GUI are still visible (and also still not updating anymore but that is expected with a disconnected device)

> 3. Run Python script to get some data from VNA. The program succed to report the number of points that I requested with ":VNA:ACQ:POINTS?" command, but got an "EMPTY" string when I requested ":VNA:TRACE:DATA? S11"

- Run the script again, this time the output is this:
```
$ python test.py
202
# GHZ S RI R 50.000000000000


202
# GHZ S RI R 50.000000000000


202
# GHZ S RI R 50.000000000000


202
# GHZ S RI R 50.000000000000
```

Okay, looks like the same problem you are reporting. Now lets look at your script in more detail:
The beginning is just the setup of the connection and some functions to read/write data, nothing unusual. Then you set up the sweep:
```python
# Set up the sweep parameters
send_scpi_command(":DEV:MODE VNA")
send_scpi_command(":VNA:SWEEP FREQUENCY")
send_scpi_command(":VNA:ACQ:IFBW 10000")
send_scpi_command(":VNA:ACQ:AVG 1")
send_scpi_command(":VNA:ACQ:POINTS 201")
send_scpi_command(":VNA:FREQuency:START 4000000000")
send_scpi_command(":VNA:FREQuency:STOP 6000000000")
```
All good, the sweep is configured now. Next you want to capture four consecutive sweeps and print the data from them:
```python
# Sweeps loop
for n in range(4):
    send_scpi_command(":VNA:ACQuisition:SINGLE TRUE")

    # Wait for the sweep to finish
    while query_scpi_command(":VNA:ACQ:FIN?") == "FALSE":
        time.sleep(0.1)

    # Grab the data of trace S11
    npts = int(query_scpi_command(":VNA:ACQ:POINTS?")) + 1
    tmp = query_scpi_command(":VNA:TRACE:TOUCHSTONE? S11")
    print(npts)
    print(tmp)
```

You start the sweeps by enabling single sweeps. The programming guide has this to say about it:
> There are two ways to trigger a new sweep:
• Change any sweep setting (e.g. center frequency)
• Issue the command again (i.e. VNA:ACQ:SINGLE TRUE always triggers a new sweep)

Okay, setting VNA:ACQ:SINGLE to TRUE is a valid way to trigger a new sweep.

After triggering a new sweep, you wait for it to finish and then grab the data. Everything is working and you also report that you are getting the data in your step 1.

Then you perform step 2. You **disconnect** the device. You mention that you do that
> to be able to work with my script

I can't quite follow the logic here. You just used your script successfully in step 1 while the device was connected. And now you think you have to disconnect the device to use your script? But okay, sure, you can disconnect your device at this point if you want to.

And now comes step 3.
You run your script.
While the device is disconnected.
And your script triggers a new sweep four times.
And wants to read the data from each of these new sweeps.
Without a connected device.
And you are wondering why you are not getting any data anymore.
While the traces in the GUI also disappeared.

Well, yes:
### if you want to get data from the device, please leave it connected!
Alternatively you can disconnect it and still read the old data but in this case please do not trigger a new sweep (which will clear all old data first)

Honestly, I am struggling to understand the logic here. And yes, I get that there is a learning curve to the software and you are new to it. But helping here has been soooooo annoying for me. Not only has the calibration nothing to do with the problem at all (which really pointed me in the wrong direction). I also had to pull every single fact out of you by asking multiple times.

And all that time you did not mention a single time that you are trying to run your script while the device is disconnect (which is the whole problem. How do you expect to get new data?)

So please excuse the sarcasm and overall annoyed tone in this reply. That is not my usual way of handling questions. And please also do not hesitate to create new discussions/issues in the future. The LibreVNA is a passion project of mine and I love helping people use it and the feedback I get is also very helpful for improving the software.

But also understand that I just spent a significant amount of time trying to understand your problem and getting all the information. This honestly feels as if you intentionally kept the one crucial detail from me that would explain everything immediately (I know it wasn't intentional but it sure feels like it was).

For future problems: Give me a detailed step-by-step guide on how to reproduce the problem in your first post. With way, way more details than here. It is my job to help with these things. It is your job to provide clear and unambiguous descriptions of what you are doing.

**Comment by iliavits** [comment]
Thanks, Jan.
Really appreciate your efforts. It looks like a great product and I believe it will be very meaningful to us.
I didn't realize that I should stay connected through the GUI while I am communicating with the device from another program. In all my previous experience I had to release the port on one app before using it another. Not realizing this led to all the confusion. Maybe after the calibration I didn't disconnect and this is it was working. Anyway, thanks again and I will definitely get back to you with some more issues that we will encounter because this device will be used by us heavily.
In the future I promise I will try to be more strict in my descriptions, believe me I didn't have a slightest intention to mislead you :)

**Comment by jankae** [comment]
Sure, please report if you find other problems. Maybe this helps with clearing up some confusion: your script is never communicating with the LibreVNA directly. It is communicating with the GUI and the GUI in turn is communicating with the device. So as soon as you disconnect the device in the GUI, that connection is gone. Your script can still communicate with the GUI but if you ask the GUI to start a new sweep, it can't do that anymore.
